### PR TITLE
The Shakeship

### PR DESCRIPTION
**Pull request by Tsunamico**
# About the pull request

Admin tool to shake the ship, and a proc that shakes it too when called.

# Explain why it's good for the game

Good tool for events, and shortens some coding, along with future systems that would cause this shaking.

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

:cl:
admin: panel that allows admins to trash the ship about.
code: adds a shakeship proc that can be called to rock the shipmap.
/:cl:

### CODE REVIEW

**Comment by Tsunamico** [comment]
baseline coding is done, it is perfectly functional and sounds nice! Great for emulating an asteroid field or anythign of the likes that shakes the ship!

But now I'll be working on making the main shaking process more streamlined by making it a global var so everybody can shake the 'ol Almayer with their code!

You BETTER make Ship vs Ship mechanics with this.

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 1]
```diff
@@ -0,0 +1,55 @@
+/client/proc/shakeshipverb()
+	set name = "Shake shipmap"
+	set category = "Admin.Events"
+
+	var/client/C = usr.client
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:5]: unnecessary, client here is src (it's a client proc)

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+	var/logckey = C.ckey
+	var/drop = 0
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:7]: if this is a boolea it should be FALSE

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 3]
```diff
@@ -0,0 +1,55 @@
+/client/proc/shakeshipverb()
+	set name = "Shake shipmap"
+	set category = "Admin.Events"
+
+	var/client/C = usr.client
+	var/logckey = C.ckey
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:6]: just ckey, so this var is redundant and you can just acess it later

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+	var/sstrength = tgui_input_number(src, "How Strong?", "Don't go overboard.", 0, 10)
+	if(!sstrength)
+		return
+	var/stime = tgui_input_number(src, "Time Between Shakes?", "Don't make it too long", 0, 30)
+	if(!sstrength)
+		return
+
+	var/prompt = alert(C, "Drop people?", "Yeet?" ,"Yes","No")
+	if(prompt == "Yes")
+		drop = 1
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:22]: boolean so TRUE

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+	prompt = alert(C, "delay it?", "Make them quiver!" ,"Yes","No")
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:24]: C can just be src

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	if(prompt == "Yes")
+		delayed = TRUE
+		delayt = tgui_input_number(src, "How much delay?", "60 secs maximum", 0, 60, 0)
+		if(!delayt)
+			return
+		prompt = alert(C, "alert people?", "ARES announcement!" ,"Yes","No")
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:30]: ditto

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	else
+		log_admin("[logckey] rocked the ship! with the strength of [sstrength], and duration of [stime]")
+		message_admins("[logckey] rocked the ship! with the strength of [sstrength], and duration of [stime]")
+		if(delayed)
+			if(announce)
+				if(sstrength <= 5)
+					shipwide_ai_announcement(whattoannounce, MAIN_AI_SYSTEM, 'sound/effects/alert.ogg')
+				if(sstrength > 5)
+					shipwide_ai_announcement(whattoannounce, MAIN_AI_SYSTEM, 'sound/effects/ob_alert.ogg')
+		 sleep(delayt * 10)
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:52]: nuke this... and just use a timer

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/admin/verbs/shakeshipverb.dm (hunk 7 above)
Comment (Tsunamico) [reply code/modules/admin/verbs/shakeshipverb.dm:52]: Ah, forgot to put it as timer, whole friggin' point of having it as two separate things, soz.

**Comment by Tsunamico** [comment]
Forgot to review some sounds that can sound better, open to suggestions!

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+	if(prompt == "Yes")
+		drop = TRUE
+
+	prompt = alert(usr.client, "delay it?", "Make them quiver!" ,"Yes","No")
+	if(prompt == "Yes")
+		delayed = TRUE
+		delayt = tgui_input_number(src, "How much delay?", "60 secs maximum", 0, 60, 0)
+		if(!delayt)
+			return
+		prompt = alert(usr.client, "alert people?", "ARES announcement!" ,"Yes","No")
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:28]: send to src

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+		return
+	var/stime = tgui_input_number(src, "Time Between Shakes?", "Don't make it too long", 0, 30)
+	if(!sstrength)
+		return
+
+	var/prompt = alert(usr.client, "Drop people?", "Yeet?" ,"Yes","No")
+	if(prompt == "Yes")
+		drop = TRUE
+
+	prompt = alert(usr.client, "delay it?", "Make them quiver!" ,"Yes","No")
```
[the remark this answers is not in the document]
Comment (harryob) [reply code/modules/admin/verbs/shakeshipverb.dm:22]: + capitals

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 10]
```diff
@@ -0,0 +1,54 @@
+/client/proc/shakeshipverb()
+	set name = "Shake shipmap"
+	set category = "Admin.Events"
+
+	var/drop = FALSE
+	var/delayed
+	var/announce
+	var/delayt = 1
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:8]: 0.1 SECONDS?

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 11]
```diff
@@ -0,0 +1,54 @@
+/client/proc/shakeshipverb()
+	set name = "Shake shipmap"
+	set category = "Admin.Events"
+
+	var/drop = FALSE
+	var/delayed
+	var/announce
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:7]: put these above where they're used

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 12]
```diff
@@ -0,0 +1,54 @@
+/client/proc/shakeshipverb()
+	set name = "Shake shipmap"
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:2]: caps

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+		if(prompt == "Yes")
+			announce = TRUE
+			whattoannounce = input(usr, "Please enter announcement text. Keep it empty to keep the default.", "What?", "")
+			if(!whattoannounce)
+				if(sstrength <= 7)
+					whattoannounce = "WARNING, IMPACT IMMINENT. ETA: [delayt] SECONDS. BRACE BRACE BRACE."
+				if(sstrength > 7)
+					whattoannounce = "DANGER, DANGER! HIGH ENERGY IMPACT IMMINENT. ETA: [delayt] SECONDS. BRACE BRACE BRACE."
+
+	prompt = alert(src, "Are you sure you want to shake the shipmap?", "Rock the ship!" ,"Yes","No")
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:38]: should be a tgui alert

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+	else
+		log_admin("[src] rocked the ship! with the strength of [sstrength], and duration of [stime]")
+		message_admins("[src] rocked the ship! with the strength of [sstrength], and duration of [stime]")
+		if(delayed)
+			if(announce)
+				if(sstrength <= 5)
+					shipwide_ai_announcement(whattoannounce, MAIN_AI_SYSTEM, 'sound/effects/alert.ogg')
+				if(sstrength > 5)
+					shipwide_ai_announcement(whattoannounce, MAIN_AI_SYSTEM, 'sound/effects/ob_alert.ogg')
+		 //sleep(delayt * 10)
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:50]: no commented code

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+					whattoannounce = "WARNING, IMPACT IMMINENT. ETA: [delayt] SECONDS. BRACE BRACE BRACE."
+				if(sstrength > 7)
+					whattoannounce = "DANGER, DANGER! HIGH ENERGY IMPACT IMMINENT. ETA: [delayt] SECONDS. BRACE BRACE BRACE."
+
+	prompt = alert(src, "Are you sure you want to shake the shipmap?", "Rock the ship!" ,"Yes","No")
+	if(prompt != "Yes")
+		return
+	else
+		log_admin("[src] rocked the ship! with the strength of [sstrength], and duration of [stime]")
+		message_admins("[src] rocked the ship! with the strength of [sstrength], and duration of [stime]")
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:43]: key_name_admin(src)

File: code/modules/admin/verbs/shakeshipverb.dm [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+		if(delayed)
+			if(announce)
+				if(sstrength <= 5)
+					shipwide_ai_announcement(whattoannounce, MAIN_AI_SYSTEM, 'sound/effects/alert.ogg')
+				if(sstrength > 5)
+					shipwide_ai_announcement(whattoannounce, MAIN_AI_SYSTEM, 'sound/effects/ob_alert.ogg')
+		 //sleep(delayt * 10)
+		 addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(shakeship), sstrength, stime, drop), delayt * 10)
```
Comment (harryob) [inline code/modules/admin/verbs/shakeshipverb.dm:51]: weird indentation?

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/modules/admin/verbs/shakeshipverb.dm (hunk 13 above)
Comment (Tsunamico) [reply code/modules/admin/verbs/shakeshipverb.dm:38]: Huh, please teach me why? Have seen standard prompts being used extensively and I use these normal prompts two other times here too.

**Review by Tsunamico — COMMENTED** [review]
Right, it feels like it should be golden now! Hope it is a fun event tool!

File: code/modules/almayer/shakeship.dm [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+ * despite being in /almayer, this thing should work in any shipmap.
+ * Arguments:
+ * * sstrength - How hard the camera shakes, and how hard the rocking on the ship is
+ * * stime - For how long the camera shakes
+ * * drop - If the shaking can make people fall, or be thrown away if sstrength is above 7
+ * * osound - not included in the admin verb! Can be set to FALSE if you don't want it's original sounds to play in case you want something different
+ */
+/proc/shakeship(sstrength, stime, drop, osound = TRUE)
+
+
```
Comment (harryob) [inline code/modules/almayer/shakeship.dm:16]:
```suggestion
```

File: code/modules/almayer/shakeship.dm [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+				addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(playsound_client), current_mob.client, 'sound/effects/doorcreaky.ogg', 20), 1 SECONDS)
+		if(sstrength > 7)
+			if(current_mob.client && osound)
+				playsound_client(current_mob.client, 'sound/effects/metal_crash.ogg', 100 )
+				playsound_client(current_mob.client, 'sound/effects/bigboom3.ogg', 100)
+
+			if(drop)
+				INVOKE_ASYNC(current_mob,  TYPE_PROC_REF(/atom/movable, throw_atom), get_ranged_target_turf(current_mob, pick(cardinal), sstrength-5), pick(cardinal), sstrength)
+
+			to_chat(current_mob, SPAN_HIGHDANGER("YOU ARE THROWN AROUND WITH VIOLENCE AND HIT THE DECK FULL FORCE!!"))
```
Comment (harryob) [inline code/modules/almayer/shakeship.dm:42]:
```suggestion
			to_chat(current_mob, SPAN_HIGHDANGER("YOU ARE THROWN AROUND VIOLENTLY AND HIT THE DECK WITH FULL FORCE!!"))
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by harryob — APPROVED** [review]

**Pull request merged**